### PR TITLE
Registration store fixes

### PR DESCRIPTION
**Pull request by msangoi**
The RegistrationStore should be responsible for removing existing observations (same path/registrationId) when a new observation is persisted.

### CODE REVIEW

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/InMemoryRegistrationStore.java [hunk 1]
```diff
@@ -184,10 +185,34 @@ public Deregistration removeRegistration(String registrationId) {
 
     /* *************** Leshan Observation API **************** */
 
+    /*
+     * The observation is not persisted here, it is done by the Californium layer (in the implementation of the
+     * org.eclipse.californium.core.observe.ObservationStore#add method)
+     * 
+     * (non-Javadoc)
```
Comment (jvermillard) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/InMemoryRegistrationStore.java]: (non-javadoc) :cry:

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/InMemoryRegistrationStore.java [hunk 2]
```diff
@@ -184,10 +185,34 @@ public Deregistration removeRegistration(String registrationId) {
...
     @Override
-    public Observation addObservation(String registrationId, Observation observation) {
-        // not used observation was added by Californium
-        return null;
+    public Collection<Observation> addObservation(String registrationId, Observation observation) {
+
+        List<Observation> removed = new ArrayList<>();
+
+        // cancel existing observations for the same path and registration id.
+        for (Observation obs : getObservations(registrationId)) {
```
Comment (sbernard31) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/InMemoryRegistrationStore.java]: I think we should move the `for` statement in the `writeLock` block

File: leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java [hunk 3]
```diff
@@ -324,10 +324,56 @@ private Registration deserializeReg(byte[] data) {
...
     @Override
-    public Observation addObservation(String registrationId, Observation observation) {
-        // not used observation was added by Californium
-        return null;
+    public Collection<Observation> addObservation(String registrationId, Observation observation) {
+
+        List<Observation> removed = new ArrayList<>();
+
+        // cancel existing observations for the same path and registration id.
+        for (Observation obs : getObservations(registrationId)) {
```
Comment (sbernard31) [inline leshan-server-cluster/src/main/java/org/eclipse/leshan/server/cluster/RedisRegistrationStore.java]: same thing here, moving the `for` statement in the `RedisLock` block

**Comment by sbernard31** [comment]
Integrated in master (commit 43f41822f7757d94b89e44371c7101fb150887bb, ff2b1c817c48f95be11701c503647c1bc62e0ca7)

**Pull request closed without merging**